lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -1116,6 +1117,253 @@ static bool canonicalizeOrOfConcatsWithCstOperands(OrOp op, size_t concatIdx1,
+/// an interval is emitted only when at least 3 equality checks are folded into
+/// two unsigned comparisons.
+///
+/// All constants and operations operate in unsigned mode.
+///
+static bool tryMergeRanges(OrOp op, PatternRewriter &rewriter) {
+  // x == value
+  struct EqCheck {
+    APInt value;
+    Value check;
+  };
+  // start < x < end
+  struct RangeCheck {
+    APInt start;
+    APInt end;
+    Value startCheck;
+    Value endCheck;
+  };
+  // x < value
+  struct BoundCheck {
+    APInt value;
+    Value check;
+  };
+
+  using AnyCheck = std::variant<RangeCheck, EqCheck, BoundCheck>;
+
+  // Identify all the relevant patterns among the inputs,
+  // mark all others to be retained unchanged.
+  auto inputs = op.inputs();
+  unsigned numKeptOperands = 0;
+  std::vector<bool> keptOperands(inputs.size());
+  DenseMap<Value, SmallVector<std::pair<unsigned, AnyCheck>>> argChecks;
+
+  for (unsigned i = 0, n = inputs.size(); i < n; ++i) {
+    auto input = inputs[i];
+
+    // Find bound checks: and(x > n, x < m)
+    auto andOp = input.getDefiningOp<AndOp>();
+    if (andOp && andOp.inputs().size() == 2) {
+      ICmpOp lhsOp = andOp.inputs()[0].getDefiningOp<ICmpOp>();
+      ICmpOp rhsOp = andOp.inputs()[1].getDefiningOp<ICmpOp>();
+
+      if (lhsOp && rhsOp && lhsOp.lhs() == rhsOp.lhs()) {
+        Value arg = lhsOp.lhs();
+
+        APInt lhsBound, rhsBound;
+        if (matchPattern(lhsOp.rhs(), m_RConstant(lhsBound)) &&
+            matchPattern(rhsOp.rhs(), m_RConstant(rhsBound))) {
+          ICmpPredicate lhsPred = lhsOp.predicate();
+          ICmpPredicate rhsPred = rhsOp.predicate();
+          if (lhsPred == ICmpPredicate::ugt && rhsPred == ICmpPredicate::ult) {
+            argChecks[arg].emplace_back(i,
+                RangeCheck{lhsBound, rhsBound, lhsOp, rhsOp});
+            continue;
+          }
+          if (lhsPred == ICmpPredicate::ult && rhsPred == ICmpPredicate::ugt) {
+            argChecks[arg].emplace_back(i,
+                RangeCheck{rhsBound, lhsBound, rhsOp, lhsOp});
+            continue;
+          }
+        }
+      }
+    }
+
+    if (auto cmpOp = input.getDefiningOp<ICmpOp>()) {
+      APInt value;
+      if (matchPattern(cmpOp.rhs(), m_RConstant(value))) {
+        // Find equality tests: x == n
+        if (cmpOp.predicate() == ICmpPredicate::eq) {
+          argChecks[cmpOp.lhs()].emplace_back(i, EqCheck{value, cmpOp});
+          continue;
+        }
+        // Find upper bound tests: x < n
+        if (cmpOp.predicate() == ICmpPredicate::ult) {
+          argChecks[cmpOp.lhs()].emplace_back(i, BoundCheck{value, cmpOp});
+          continue;
+        }
+      }
+    }
+
+    keptOperands[i] = true;
+    ++numKeptOperands;
+  }
+
+  // Find an 'n' such that 'n < x' for any 'x' matching the check. If 0 is part
+  // of any of the checks, nullopt is returned to represent -1.
+  auto getLowerBound = [](const AnyCheck &check) -> std::optional<APInt> {
```

Comment:
Maybe the AnyCheck could be a small helper class that knows how to pick between the variants, and these functions could become member functions? I'm thinking of something like this, for example: https://github.com/llvm/circt/blob/a2478bfd265cd3d3771c196b9ed0acbb46b3e20f/lib/Conversion/SCFToCalyx/SCFToCalyx.cpp#L48